kernel/src/engine/arrow_utils.rs

```diff
@@ -495,6 +504,89 @@ fn get_indices(
     ))
 }
 
+type FieldIndex = usize;
+
+/// contains information about a StructField matched to a parquet struct field
+///
+/// # Lifetime Parameters
+/// * `'k` - The lifetime of the referenced kernel StructField
+struct KernelFieldInfo<'k> {
+    /// The index of the struct field in its parent struct
+    parquet_index: FieldIndex,
+    /// A reference to the struct field
+    field: &'k StructField,
+}
+
+/// Contains a information about a parquet field and the matching `KernelFieldInfo` if one
+/// exists. Parquet struct fields are matched to Kernel fields in [`match_parquet_fields`].
+///
+/// # Lifetime Parameters
+/// * `'k` - The lifetime of the referenced kernel StructField
+/// * `'p` - The lifetime of the referenced parquet ArrowField
+struct MatchedParquetField<'p, 'k> {
+    /// The index of the parquet field
+    parquet_index: FieldIndex,
+    /// A reference to the parquet field in the arrow schema
+    parquet_field: &'p ArrowField,
+    /// If present, this is a `KernelFieldInfo` belonging to a matching kernel `StructField`
+    kernel_field_info: Option<KernelFieldInfo<'k>>,
+}
+
+/// Constructs an iterator where each parquet Field in `fields` is matched
+/// with a a kernel `KernelFieldInfo` representing a StructField.
+///
+/// The iterator returned has a [`MatchedParquetField`] for each element in `parquet_fields`.
+fn match_parquet_fields<'k, 'p>(
+    kernel_schema: &'k StructType,
+    parquet_fields: &'p ArrowFields,
+) -> impl Iterator<Item = MatchedParquetField<'p, 'k>> {
+    type FieldId = i64;
+    // Construct a map from the field id to its StructField name.
+    let field_id_to_name: HashMap<FieldId, &String> = kernel_schema
```

Comment:
This is only used if we find field ids, could we use a `LazyLock` or `OnceLock` to make its computation lazy? I don't think we want to pay the cost of computing this whole map when it's not needed.

Reply:
Cool thanks. I realize now it could probably even just be an `Option` that you fill in since I don't think we expect this to be called in a multi-threaded context. `OnceLock` is fine too though.